File: services/api/src/tadas/services/api/routers/tenancy.py
"""Tenancy routes: sign in through the identity provider, choose or switch a
tenant, read the principal, create an org, manage members and their
invitations, open single sign-on, sessions, and api keys. Every function is
one call into the tenancy service."""

from uuid import UUID

from fastapi import APIRouter, Depends, Response

from tadas.services.api.gateway.auth import Ctx, Identity, Rctx, dev_sign_in_open
from tadas.services.api.gateway.idempotency import Idem
from tadas.services.api.gateway.ratelimit import rate_limited
from tadas.services.api.gateway.resolve import TenancyService
from tadas.services.api.types.common import LIMIT_DEFAULT
from tadas.services.api.types.tenancy import (
    AddApiKeyRequest,
    ApiKeyPageView,
    ApiKeyView,
    CreateTeamOrgRequest,
    DeviceSignInView,
    DeviceTokenRequest,
    DevSignInRequest,
    ExchangeSessionRequest,
    IdentityView,
    InvitationPageView,
    InvitationView,
    InviteMemberRequest,
    IssuedApiKeyView,
    IssuedLoginView,
    IssuedSessionView,
    LogoutRequest,
    MembershipChoicePageView,
    MembershipChoiceView,
    MembershipPageView,
    MembershipView,
    MeView,
    OrgView,
    SecondFactorRequest,
    SessionView,
    SignedOutView,
    SignInCallbackRequest,
    SignInStartRequest,
    SignInStartView,
    SsoLinkRequest,
    SsoLinkView,
    UpdateIdentityRequest,
    UpdateMembershipRequest,
    UpdateMeRequest,
    UserPageView,
    UserView,
)

router = APIRouter(tags=["tenancy"])


# The sign-in routes carry no Idempotency-Key: the marker is kept per tenant
# and principal, and none exists yet. Each is counted against the
# per-address sign-in budget. A code is exchanged once, so a retry after a
# lost answer is refused and the person signs in again.
@router.post("/auth/sign-in", response_model=SignInStartView, dependencies=[rate_limited("login")])
async def start_sign_in(
    rctx: Rctx, tenancy: TenancyService, body: SignInStartRequest
) -> SignInStartView:
    return await tenancy.start_sign_in(rctx, body)


@router.post("/auth/callback", response_model=IssuedLoginView, dependencies=[rate_limited("login")])
async def finish_sign_in(
    rctx: Rctx, tenancy: TenancyService, body: SignInCallbackRequest
) -> IssuedLoginView:
    return await tenancy.finish_sign_in(rctx, body)


@router.post("/auth/device", response_model=DeviceSignInView, dependencies=[rate_limited("login")])
async def start_device_sign_in(rctx: Rctx, tenancy: TenancyService) -> DeviceSignInView:
    return await tenancy.start_device_sign_in(rctx)


@router.post(
    "/auth/device/token", response_model=IssuedLoginView, dependencies=[rate_limited("login")]
)
async def finish_device_sign_in(
    rctx: Rctx, tenancy: TenancyService, body: DeviceTokenRequest
) -> IssuedLoginView:
    return await tenancy.finish_device_sign_in(rctx, body)


# Local and test only: 404 wherever it is off, as a route that does not
# exist answers, and a deployed process refuses to start with it on.
@router.post(
    "/auth/dev-sign-in",
    response_model=IssuedLoginView,
    dependencies=[Depends(dev_sign_in_open), rate_limited("login")],
)
async def dev_sign_in(
    rctx: Rctx, tenancy: TenancyService, body: DevSignInRequest
) -> IssuedLoginView:
    return await tenancy.dev_sign_in(rctx, body)


@router.post(
    "/auth/second-factor", response_model=IssuedLoginView, dependencies=[rate_limited("login")]
)
async def verify_second_factor(
    identity: Identity, tenancy: TenancyService, body: SecondFactorRequest
) -> IssuedLoginView:
    return await tenancy.verify_second_factor(identity, body)


@router.post("/auth/sessions", response_model=IssuedSessionView)
async def exchange_session(
    identity: Identity, tenancy: TenancyService, body: ExchangeSessionRequest
) -> IssuedSessionView:
    return await tenancy.exchange_session(identity, body)


@router.get("/auth/memberships", response_model=MembershipChoicePageView)
async def list_my_memberships(
    identity: Identity,
    tenancy: TenancyService,
    cursor: str | None = None,
    limit: int = LIMIT_DEFAULT,
) -> MembershipChoicePageView:
    return await tenancy.get_identity_memberships(identity, cursor, limit)


# The session presented ends; the answer names where the browser goes to end
# the identity provider's session behind it, when there is one. The body is
# optional: a caller with no browser sends none.
@router.post("/auth/logout", response_model=SignedOutView)
async def logout(
    ctx: Ctx, tenancy: TenancyService, body: LogoutRequest | None = None
) -> SignedOutView:
    return await tenancy.logout(ctx, body)


@router.get("/me", response_model=MeView)
async def me(ctx: Ctx, tenancy: TenancyService) -> MeView:
    return await tenancy.get_me(ctx)


@router.patch("/me", response_model=UserView)
async def update_me(ctx: Ctx, tenancy: TenancyService, body: UpdateMeRequest) -> UserView:
    return await tenancy.update_me(ctx, body)


@router.get("/me/identity", response_model=IdentityView)
async def my_identity(ctx: Ctx, tenancy: TenancyService) -> IdentityView:
    return await tenancy.get_identity(ctx)


@router.patch("/me/identity", response_model=IdentityView)
async def update_my_identity(
    ctx: Ctx, tenancy: TenancyService, body: UpdateIdentityRequest
) -> IdentityView:
    return await tenancy.update_identity(ctx, body)


@router.get("/orgs/current", response_model=OrgView)
async def current_org(ctx: Ctx, tenancy: TenancyService) -> OrgView:
    return await tenancy.get_org(ctx)


# A team org the caller makes and owns, answered with the caller's place in
# it; the switch there is the exchange. The idempotency marker is the
# caller's in the tenant they call from, and the org is created on its id.
@router.post("/orgs", response_model=MembershipChoiceView, status_code=201)
async def create_org(
    ctx: Ctx, tenancy: TenancyService, body: CreateTeamOrgRequest, idem: Idem
) -> Response:
    return await idem.run(201, lambda attempt: tenancy.create_org(ctx, body, attempt))


@router.get("/users", response_model=UserPageView)
async def list_users(
    ctx: Ctx, tenancy: TenancyService, cursor: str | None = None, limit: int = LIMIT_DEFAULT
) -> UserPageView:
    return await tenancy.get_users(ctx, cursor, limit)


@router.get("/memberships", response_model=MembershipPageView)
async def list_memberships(
    ctx: Ctx, tenancy: TenancyService, cursor: str | None = None, limit: int = LIMIT_DEFAULT
) -> MembershipPageView:
    return await tenancy.get_memberships(ctx, cursor, limit)


@router.patch("/memberships/{user_id}", response_model=MembershipView)
async def update_membership(
    ctx: Ctx, tenancy: TenancyService, user_id: UUID, body: UpdateMembershipRequest
) -> MembershipView:
    return await tenancy.update_membership_role(ctx, user_id, body)


@router.delete("/memberships/{user_id}", response_model=UserView)
async def remove_member(ctx: Ctx, tenancy: TenancyService, user_id: UUID) -> UserView:
    return await tenancy.remove_member(ctx, user_id)


@router.get("/invitations", response_model=InvitationPageView)
async def list_invitations(
    ctx: Ctx, tenancy: TenancyService, cursor: str | None = None, limit: int = LIMIT_DEFAULT
) -> InvitationPageView:
    return await tenancy.get_invitations(ctx, cursor, limit)


# The identity provider sends the email. The idempotency marker is the
# caller's, and the invitation is created on its id, so a retry sends one.
@router.post("/invitations", response_model=InvitationView, status_code=201)
async def invite_member(
    ctx: Ctx, tenancy: TenancyService, body: InviteMemberRequest, idem: Idem
) -> Response:
    return await idem.run(201, lambda attempt: tenancy.invite_member(ctx, body, attempt))


@router.post("/invitations/{invitation_id}/resend", response_model=InvitationView)
async def resend_invitation(
    ctx: Ctx, tenancy: TenancyService, invitation_id: UUID
) -> InvitationView:
    return await tenancy.resend_invitation(ctx, invitation_id)


@router.delete("/invitations/{invitation_id}", response_model=InvitationView)
async def revoke_invitation(
    ctx: Ctx, tenancy: TenancyService, invitation_id: UUID
) -> InvitationView:
    return await tenancy.revoke_invitation(ctx, invitation_id)


@router.post("/orgs/current/sso-link", response_model=SsoLinkView)
async def sso_link(ctx: Ctx, tenancy: TenancyService, body: SsoLinkRequest) -> SsoLinkView:
    return await tenancy.sso_link(ctx, body)


@router.get("/sessions", response_model=list[SessionView])
async def list_sessions(
    ctx: Ctx, tenancy: TenancyService, limit: int = LIMIT_DEFAULT
) -> list[SessionView]:
    return await tenancy.get_sessions(ctx, limit)


@router.delete("/sessions/{session_id}", response_model=SessionView)
async def revoke_session(ctx: Ctx, tenancy: TenancyService, session_id: UUID) -> SessionView:
    return await tenancy.revoke_session(ctx, session_id)


@router.get("/api-keys", response_model=ApiKeyPageView)
async def list_api_keys(
    ctx: Ctx, tenancy: TenancyService, cursor: str | None = None, limit: int = LIMIT_DEFAULT
) -> ApiKeyPageView:
    return await tenancy.get_api_keys(ctx, cursor, limit)


@router.post("/api-keys", response_model=IssuedApiKeyView, status_code=201)
async def create_api_key(
    ctx: Ctx, tenancy: TenancyService, body: AddApiKeyRequest, idem: Idem
) -> Response:
    return await idem.run(201, lambda attempt: tenancy.create_api_key(ctx, body, attempt))


@router.delete("/api-keys/{api_key_id}", response_model=ApiKeyView)
async def revoke_api_key(ctx: Ctx, tenancy: TenancyService, api_key_id: UUID) -> ApiKeyView:
    return await tenancy.revoke_api_key(ctx, api_key_id)

File: apps/portal/src/queries/tenancy.ts
import { useInfiniteQuery, useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { useEffect } from "react";
import type {
  AccountDeletedView,
  AddApiKeyRequest,
  ApiKeyPageView,
  ApiKeyView,
  CreateTeamOrgRequest,
  DeleteAccountRequest,
  DevSignInRequest,
  ExchangeSessionRequest,
  IdentityView,
  IssuedApiKeyView,
  IssuedLoginView,
  IssuedSessionView,
  InvitationPageView,
  InvitationView,
  InviteMemberRequest,
  LogoutRequest,
  MembershipChoicePageView,
  MembershipChoiceView,
  MeView,
  SignedOutView,
  SignInCallbackRequest,
  SignInStartRequest,
  SignInStartView,
  SsoLinkRequest,
  SsoLinkView,
  UserPageView,
  UserView,
} from "../api";
import { api } from "../app/api";
import { keys } from "./keys";

/** Both lists are paged by the server's cursor; these are page sizes, not
 * ceilings, so nothing is hidden past the end of the first page. The member
 * list is read whole (below), so it asks for the largest page the server
 * gives; the key list is read a page at a time on the screen that shows it. */
const USERS_PAGE_SIZE = 200;
export const API_KEYS_PAGE_SIZE = 50;
export const INVITATIONS_PAGE_SIZE = 50;
/** A person belongs to at most a hundred orgs; the chip reads them whole. */
const MY_MEMBERSHIPS_PAGE_SIZE = 200;

export function useMe() {
  return useQuery({ queryKey: keys.me, queryFn: ({ signal }) => api.get<MeView>("/v1/me", { signal }) });
}

/** The person behind the session, across every org: the time zone lives here. */
export function useIdentity() {
  return useQuery({
    queryKey: keys.identity,
    queryFn: ({ signal }) => api.get<IdentityView>("/v1/me/identity", { signal }),
  });
}

/** Every member of the org, page after page. The list is not a screen of its
 * own: it names the people on tasks and fills the assignee picker, and a
 * member missing from it reads as a former member and cannot be assigned, so this
 * follows the cursor to the end rather than stopping at one page. The pages
 * loaded so far are returned as they arrive. */
export function useUsers() {
  const query = useInfiniteQuery({
    queryKey: keys.users.list(USERS_PAGE_SIZE),
    initialPageParam: null as string | null,
    getNextPageParam: (last: UserPageView) => last.next_cursor,
    queryFn: ({ pageParam, signal }) => {
      const cursor = pageParam ? `&cursor=${encodeURIComponent(pageParam)}` : "";
      return api.get<UserPageView>(`/v1/users?limit=${USERS_PAGE_SIZE}${cursor}`, { signal });
    },
  });
  const { hasNextPage, isFetchingNextPage, isFetchNextPageError, fetchNextPage } = query;
  // A page that failed is not asked for again on its own: retries are the
  // client's, and re-firing here on every render would be a loop against a
  // server that just refused. The pages in hand are what the app shows, and
  // the next invalidation starts the walk over.
  const walking = hasNextPage && !isFetchNextPageError;
  useEffect(() => {
    if (walking && !isFetchingNextPage) void fetchNextPage();
  }, [walking, isFetchingNextPage, fetchNextPage]);
  return {
    ...query,
    data: query.data?.pages.flatMap((page) => page.items) as UserView[] | undefined,
    /** Pending until the whole list is in hand: a caller that maps ids to
     * names would otherwise call a member still on the way a former member. */
    isPending: query.isPending || walking,
  };
}

/** Only a member who may manage keys asks for them: `GET /v1/api-keys`
 * refuses anyone else, and a refusal nobody can act on is not an error to
 * show. Disabled, the query stays pending and never fetches, so a caller
 * reads `enabled` and not `isPending` to know whether to wait. The list is
 * paged on demand: the screen shows the first page and asks for the next. */
export function useApiKeys(enabled = true) {
  const query = useInfiniteQuery({
    queryKey: keys.apiKeys.list(API_KEYS_PAGE_SIZE),
    initialPageParam: null as string | null,
    getNextPageParam: (last: ApiKeyPageView) => last.next_cursor,
    queryFn: ({ pageParam, signal }) => {
      const cursor = pageParam ? `&cursor=${encodeURIComponent(pageParam)}` : "";
      return api.get<ApiKeyPageView>(`/v1/api-keys?limit=${API_KEYS_PAGE_SIZE}${cursor}`, {
        signal,
      });
    },
    enabled,
  });
  return {
    ...query,
    data: query.data?.pages.flatMap((page) => page.items) as ApiKeyView[] | undefined,
  };
}

export function useCreateApiKey() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (body: AddApiKeyRequest) =>
      api.post<IssuedApiKeyView>("/v1/api-keys", body, { idempotencyKey: crypto.randomUUID() }),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: keys.apiKeys.all }),
  });
}

export function useRevokeApiKey() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (apiKeyId: string) => api.del<ApiKeyView>(`/v1/api-keys/${apiKeyId}`),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: keys.apiKeys.all }),
  });
}

/** The orgs the signed-in person belongs to, page after page, read with the
 * session this tab holds. The chip needs the whole list to offer a switch. */
export function useMyMemberships() {
  const query = useInfiniteQuery({
    queryKey: keys.myMemberships.list(MY_MEMBERSHIPS_PAGE_SIZE),
    initialPageParam: null as string | null,
    getNextPageParam: (last: MembershipChoicePageView) => last.next_cursor,
    queryFn: ({ pageParam, signal }) => {
      const cursor = pageParam ? `&cursor=${encodeURIComponent(pageParam)}` : "";
      return api.get<MembershipChoicePageView>(
        `/v1/auth/memberships?limit=${MY_MEMBERSHIPS_PAGE_SIZE}${cursor}`,
        { signal },
      );
    },
  });
  const { hasNextPage, isFetchingNextPage, isFetchNextPageError, fetchNextPage } = query;
  const walking = hasNextPage && !isFetchNextPageError;
  useEffect(() => {
    if (walking && !isFetchingNextPage) void fetchNextPage();
  }, [walking, isFetchingNextPage, fetchNextPage]);
  return {
    ...query,
    data: query.data?.pages.flatMap((page) => page.items) as MembershipChoiceView[] | undefined,
    isPending: query.isPending || walking,
  };
}

/** Where the browser goes to sign in at the identity provider. Sent once,
 * with no bearer: nothing is signed in yet. */
export function useStartSignIn() {
  return useMutation({
    mutationFn: (body: SignInStartRequest) =>
      api.post<SignInStartView>("/v1/auth/sign-in", body, { token: null }),
  });
}

/** The code the browser brought back, exchanged by the API. A code is good
 * once, so this is sent once; the answer is a sign-in, and a person nobody
 * knew is signed up by it, with their personal org. */
export function useFinishSignIn() {
  return useMutation({
    mutationFn: (body: SignInCallbackRequest) =>
      api.post<IssuedLoginView>("/v1/auth/callback", body, { token: null }),
  });
}

/** The local stack's sign-in by address alone; a deployed API answers 404. */
export function useDevSignIn() {
  return useMutation({
    mutationFn: (body: DevSignInRequest) =>
      api.post<IssuedLoginView>("/v1/auth/dev-sign-in", body, { token: null }),
  });
}

/** Revokes the session the token names. The server answers with the session,
 * revoked, and where the browser goes to end the identity provider's session
 * behind it, which sends the browser back to `return_to`. */
export function useLogout() {
  return useMutation({
    mutationFn: (body: LogoutRequest) => api.post<SignedOutView>("/v1/auth/logout", body),
  });
}

/** Deletes the caller's whole account. The server answers once it is gone,
 * with where the browser goes to end the identity provider's session, as a
 * sign-out does. Sent once: a retry would meet a session that is gone. */
export function useDeleteAccount() {
  return useMutation({
    mutationFn: (body: DeleteAccountRequest) => api.post<AccountDeletedView>("/v1/me/deletion", body),
  });
}

export function useExchangeSession() {
  return useMutation({
    mutationFn: ({ loginToken, body }: { loginToken: string; body: ExchangeSessionRequest }) =>
      api.post<IssuedSessionView>("/v1/auth/sessions", body, { token: loginToken }),
  });
}

/** The switch: the exchange presented with the session this tab holds. The
 * server ends that session in the same write and answers with the new one. */
export function useSwitchOrg() {
  return useMutation({
    mutationFn: (orgId: string) => api.post<IssuedSessionView>("/v1/auth/sessions", { org_id: orgId }),
  });
}

/** A team org the signed-in person makes and owns, from the tenant they are
 * in. A creating POST, so it carries an idempotency key and a retry makes
 * one org. The answer is the person's place in it, which the switch takes;
 * the list of places is read again, since no push names it. */
export function useCreateOrg() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (body: CreateTeamOrgRequest) =>
      api.post<MembershipChoiceView>("/v1/orgs", body, { idempotencyKey: crypto.randomUUID() }),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: keys.myMemberships.all }),
  });
}

/** The org's pending invitations, for a member who manages members; paged on
 * demand like the key list. Disabled for anyone else, who would be refused. */
export function useInvitations(enabled = true) {
  const query = useInfiniteQuery({
    queryKey: keys.invitations.list(INVITATIONS_PAGE_SIZE),
    initialPageParam: null as string | null,
    getNextPageParam: (last: InvitationPageView) => last.next_cursor,
    queryFn: ({ pageParam, signal }) => {
      const cursor = pageParam ? `&cursor=${encodeURIComponent(pageParam)}` : "";
      return api.get<InvitationPageView>(`/v1/invitations?limit=${INVITATIONS_PAGE_SIZE}${cursor}`, {
        signal,
      });
    },
    enabled,
  });
  return {
    ...query,
    data: query.data?.pages.flatMap((page) => page.items) as InvitationView[] | undefined,
  };
}

/** The identity provider sends the email with the link. A creating POST, so
 * it carries an idempotency key and a retry sends one invitation. */
export function useInviteMember() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (body: InviteMemberRequest) =>
      api.post<InvitationView>("/v1/invitations", body, { idempotencyKey: crypto.randomUUID() }),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: keys.invitations.all }),
  });
}

export function useResendInvitation() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (invitationId: string) => api.post<InvitationView>(`/v1/invitations/${invitationId}/resend`),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: keys.invitations.all }),
  });
}

export function useRevokeInvitation() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (invitationId: string) => api.del<InvitationView>(`/v1/invitations/${invitationId}`),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: keys.invitations.all }),
  });
}

/** A short-lived link to the identity provider's admin portal for this org. */
export function useSsoLink() {
  return useMutation({
    mutationFn: (body: SsoLinkRequest) => api.post<SsoLinkView>("/v1/orgs/current/sso-link", body),
  });
}
